### DISCUSSION TITLE
How javascript app hooks works?

### DISCUSSION THREAD

**Discussion by denisgurev** [pocketbase/pocketbase#2767] [Q&A] [answered]
I noticed the recent commits about the "jsvm" which I suppose is the server js hooks, right?

I'm starting a new project and was considering to use SvelteKit with the node adapter as a backend, but if PocketBase will have js server support I may wait for that and use just the spa adapter.

I'll not ask for eta because I know you don't like it, but can you share some details how this will work and what will be supported with js? Can I send emails with js hooks? Can I add new actions to execute other stuffs like calling Stripe?

**Comment by ganigeorgiev** [comment, marked as answer]
Yes, by "jsvm" I refer to the [goja](https://github.com/dop251/goja) plugin implementation.

> can you share some details how this will work

I haven't pushed yet everything in the [`develop` branch](https://github.com/pocketbase/pocketbase/tree/develop), but the existing code should be runnable. If you want to test what we have so far, you can build PocketBase manually:

  0. Install [Go 1.18+](https://go.dev/doc/install)
  1. `git clone --single-branch --branch develop https://github.com/pocketbase/pocketbase.git`
  2. Navigate to `examples/base`
  3. Run `CGO_ENABLED=0 go build`
  4. Start the created executable by running `./base serve`.

**You don't need to have node,deno,etc. to use the hooks.** The JavaScript engine is written in pure Go and it implements the ES5 standard.

To start using the JS hooks, you'll need only to create `*.pb.js` file(s) inside a `pb_hooks` directory. You can put all your hooks inside a single file or split it in multiple files.
There is a file watcher that will restart the application if `pb_hooks` changes (currently supported only on UNIX based systems).

The PocketBase app instance is exposed via the global `$app` variable.
Similarly there are other global helpers exposed like `$tokens`, `$apis`, `$security`, etc.
All of this will be documented.

All exposed JS APIs are semi-auto generated from the Go ones and you can do almost everything similarly to the Go hooks.

Currently there is no http client helper, but it will be added soon.

Here are some examples from the documentation I'm currently working on with both Go and JS equivalent code for comparison:

- **Bind to event hooks and intercept responses:**
  ```go
  // go
  app.OnRecordBeforeCreateRequest("posts").Add(func(e *core.RecordCreateEvent) error {
      requestData := apis.RequestData(e.HttpContext)

      // if not an admin, overwrite the newly submitted "posts" record status to pending
      if requestData.Admin == nil {
          e.Record.Set("status", "pending")
      }

      return nil
  })
  ```
  ```js
  // pb_hooks/test.pb.js
  $app.onRecordBeforeCreateRequest("posts").add((e) => {
      let requestData = $apis.requestData(e.httpContext)

      // if not an admin, overwrite the newly submitted "posts" record status to pending
      if (!requestData.admin) {
          e.record.set("status", "pending")
      }
  })
  ```

- **Register custom routes:**

  ```go
  // go
  app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
    e.Router.GET(
          "/hello",
          func(c echo.Context) error {
              return c.String(http.StatusOK, "Hello world!")
          },
          // middlewares
          apis.ActivityLogger(app),
          apis.RequireAdminAuth(),
      )

      return nil
  })
  ```
  ```js
  // pb_hooks/test.pb.js
  $app.onBeforeServe().add((e) => {
      e.router.get(
          "/hello",
          (c) => {
              return c.string(200, "Hello world!")
          },
          // middlewares
          $apis.activityLogger($app),
          $apis.requireAdminAuth(),
      )
  })
  ```

- **Register custom console commands:**
  ```go
  // go
  app.RootCmd.AddCommand(&cobra.Command{
      Use: "hello",
      Run: func(command *cobra.Command, args []string) {
          print("Hello world!")
      },
  })
  ```
  ```js
  // pb_hooks/test.pb.js
  $app.rootCmd.addCommand(new Command({
      use: "hello",
      run: (command, args) => {
          console.log("Hello world!")
      },
  })
  ```

- **Sending email**
  ```go
  // go
  message := &mailer.Message{
      From: mail.Address{
          Address: app.Settings().Meta.SenderAddress,
          Name:    app.Settings().Meta.SenderName,
      },
      To:      []mail.Address{{Address: "test@example.com"}},
      Subject: "YOUR_SUBJECT...",
      HTML:    "YOUR_HTML_BODY...",
  }

  if err := app.NewMailClient().Send(message); err != nil {
      return err
  }
  ```
  ```js
  // pb_hooks/test.pb.js
  let message = new MailerMessage({
      from: {
          address: $app.settings().meta.senderAddress,
          name:    $app.settings().meta.senderName,
      },
      to:      [{address: "test@example.com"}],
      subject: "YOUR_SUBJECT...",
      html:    "YOUR_HTML_BODY...",
  })

  app.newMailClient().send(message);
  ```

- **Custom db query**
  ```go
  // go
  type row struct {
      ColumnA string          `db:"columnA"`
      ColumnB types.JsonArray `db:"columnB"`
      Total   int             `db:"total"`
  }

  result := []*row{}

  err := app.Dao().DB().
      Select("columnA", "columnB", "sum(columnC) as total").
      From("example").
      Limit(100).
      GroupBy("columnA").
      All(&result)

  if err != nil {
      return err
  }

  print(result[0].Total)
  ```
  ```js
  // pb_hooks/test.pb.js
  let result = new DynamicList({
      // describe the shape of the data (acts also as initial values)
      "columnA": "",
      "columnB": [],
      "total":   0,
  });

  $app.dao().db().
      select("columnA", "columnB", "sum(columnC) as total").
      from("example").
      limit(100).
      groupBy("columnA").
      all(result)

  console.log(result[0].total)
  ```

Note that you can't use directly TypeScript (_unless you manually transpile it to JS yourself_), but the binary ships with ambient TypeScript declartions that will help with the codeintelisense as long as your editor has TypeScript LSP support (_VSCode I think comes bundled with it_).

You can find more information on the expected TypeScript declarations workflow in https://github.com/dop251/goja/discussions/516.

I'm still working on refining the integration and I don't want to rush it. I know that I've spend way more time on it than I initially anticipated but at the same time I want to release it with good enough set of features from the beginning so after that I can finally focus more on Presentator.

**What remains from my main targets for v0.17.0:**

- documentation that describes the new JavaScript apis, and expand the existing Go guides (_this is the most time consuming part so far_).
- http client helper for sending `get`, `post`, `delete`, etc. requests from the JS hooks
- experiment with combined before and after hooks to allow sharing data between the 2 states (_this will be my second attempt for this as the first implementation from couple weeks ago didn't work well with the db transactions, but now I think I have better idea how to handle it_).

**Reply by denisgurev** [reply]
Wow, the next release will be sick! You already covered 90% of what I need :smiley: 

I can start working on the frontend of my app then and by the time I have to touch server code I'll try to build pb myself if it is not released yet.

Thank you @ganigeorgiev :heart:

**Reply by benallfree** [reply]
@ganigeorgiev 

> http client helper for sending get, post, delete, etc. requests from the JS hooks

Thinking about this from the https://pockethost.io perspective, it would be great if the JS runtime could be sandboxed with restricted I/O like Deno provides (https://deno.com/manual@v1.34.3/basics/permissions). That's the main reason I chose Deno for the cloud functions implementation. If users will be able to achieve file system and network access through `pb_scripts`, then I think it would be too dangerous to allow on PocketHost. 

What do you think?

**Reply by ganigeorgiev** [reply]
@benallfree Yes, allowing `pb_hooks` for your use case will be dangerous. Even with Deno I don't think it will be easy to achieve since there are a lot of edge cases and users will be able to starve the shared server resourses one way or another. If you want to allow your users to execute custom server-side code, I'd recommend implementing some sort of lower level isolation for example with docker/podman containers (both have options to specify [runtime resource constraints](https://docs.docker.com/config/containers/resource_constraints/)). 

Sandboxing and limiting network/file access for now is not planned as it will complicate too much the setup and doesn't make much sense for the general PocketBase use case. The intended usage of the `pb_hooks` is to allow users to do as much things as possible, similarly to the Go hooks.

**Reply by charbs-io** [reply]
Wow! This is incredibly exciting. Going to start testing this off the dev branch.

**Reply by tigawanna** [reply]
This is beyond awesome, 
Quick question 
will the pb_hooks directory support go hooks too , would be nice to have a way to load in go hooks without extending , biggest worry about JS would be performance

**Reply by ganigeorgiev** [reply]
> will the pb_hooks directory support go hooks too

No. If you want to use Go, you'll need to have Go installed and build the project.

Go is compiled language and cannot be loaded dynamically, unless you use a Go interpreter (eg. [yaegi](https://github.com/traefik/yaegi)), but the dev experience would be worst because you'll not have the same IDE/LSP support and the performance most likely would be on par with the JS interpreter.

I haven't done any intensive goja benchamark tests yet, but I don't expect the JS hooks to have noticeable impact on the performance as the JS code is loaded only once (during startup) and for most of the cases you use the exposed bindings which call the Go code. The performance may degrade in case you write heavy computation tasks in plain JS, which is not usually the primarily use case of PocketBase and for those rare cases we can always add a Go binding as helper if it is something commonly used (eg. aes encryption, random text generator, etc.).

**Reply by tigawanna** [reply]
Noted , excited to try them out

**Reply by ganigeorgiev** [reply]
Just for info in case someone end up here - the js app hooks integration unfortunately will have to be deleted as it cannot work efficiently with how PocketBase hooks execute (for more details you can read in https://github.com/pocketbase/pocketbase/discussions/2825).

**Reply by ganigeorgiev** [reply]
_Update: the JS app hooks might be implemented with the next release with some limitations. For more details please check https://github.com/pocketbase/pocketbase/discussions/2825#discussioncomment-6376210._
